### PR TITLE
Added another model tutorial

### PR DESCRIPTION
**Pull request by matt-gardner**
I re-read Joel's tutorial before starting this one, and realized that there's more overlap than I thought.  I wasn't sure how to position this one - I didn't want to just duplicate what Joel already had, but I thought another example would be helpful.  It ended up considerably longer than I thought it might (about 4 times longer than the one that's there) - maybe this should be in the "deep dive" section?  I'm not sure the tone or the aim of this tutorial really fits in a "deep dive", though.

### CODE REVIEW

File: tutorials/getting_started/another_model_example.md [hunk 1]
```diff
@@ -0,0 +1,611 @@
...
+
+Here again the "type" key matches what we registered our `Model` as, and the rest of the parameters
+get passed to our `AcademicPaperClassifier.from_params` method.  We have parameters for
+constructing the `TextFieldEmbedder`, which says to take the out of the "tokens" `TokenIndexer` and
+embed it using fixed, pretrained glove vectors.  If I wanted to also use a CNN over character ids,
+I would need to add a `TokenCharactersIndexer` to the `TokenIndexers` in the `DatasetReader`, and
+a corresponding entry to the `TextFieldEmbedder` parameters specifying a `TokenCharactersEncoder`.
+
+The title and abstract encoders are both bi-directional LSTMs.  The actual LSTM implementation
+here is a thin wrapper around pytorch's built-in LSTMs that make them conform to the
```
Comment (DeNeutoy) [inline tutorials/getting_started/another_model_example.md]: It's not thin haha

File: tutorials/getting_started/another_model_example.md [hunk 2]
```diff
@@ -0,0 +1,611 @@
...
+Well, that file is basically identical to a file that we'll use to train the model on real data,
+so let's just look at that one instead.
+
+## Step five: train the model
+
+Our [getting started tutorial](training_and_evaluating.md) says to use `python -m allennlp.run
+train CONFIG_FILE` to train a model.  That's not going to work for us, however, because
+`allennlp.run` doesn't know about the `DatasetReader` and `Model` classes that we added to the
+registry, so it will fail to load our configuration files.  Instead, we need to write our own
+`run.py` script ([link to the fifith
```
Comment (DeNeutoy) [inline tutorials/getting_started/another_model_example.md]: fifith -> fifth

File: tutorials/getting_started/another_model_example.md [hunk 3]
```diff
@@ -0,0 +1,611 @@
...
+    def decode(self, output_dict: Dict[str, torch.Tensor]) -> Dict[str, torch.Tensor]:
+        predictions = output_dict['class_probabilities'].cpu().data.numpy()
+        argmax_indices = numpy.argmax(predictions, axis=-1)
+        labels = [self.vocab.get_token_from_index(x, namespace="labels")
+                  for x in argmax_indices]
+        output_dict['label'] = labels
+        return output_dict
+```
+
+The name here is a little misleading; originally, we were using this for doing things like viterbi
```
Comment (DeNeutoy) [inline tutorials/getting_started/another_model_example.md]: Not sure this is so helpful - SRL and Coref both do a considerable amount of work in decode.

File: tutorials/getting_started/another_model_example.md [hunk 4]
```diff
@@ -0,0 +1,611 @@
...
+        params.assert_empty(cls.__name__)
+        return cls(tokenizer=tokenizer, token_indexers=token_indexers)
+```
+
+The constructor takes these dependencies as inputs, where by default we're splitting text into
+words, and representing them as single word ids (as opposed to, say, sequences of character ids).
+To get different tokenization behavior or different word representations, you just need to pass in
+different objects when you construct the `DatasetReader`.
+
+But where do I construct the `DatasetReader`, you ask?  Well, you don't, at least not if you're
```
Comment (DeNeutoy) [inline tutorials/getting_started/another_model_example.md]: This rhetorical style is used quite a lot - it's a bit much for a tutorial, maybe.

File: tutorials/getting_started/another_model_example.md [hunk 5]
```diff
@@ -0,0 +1,611 @@
...
+lines of code to get a very flexible classifier for academic papers.
+
+## First step: install AllenNLP
+
+In our [first
+commit](https://github.com/allenai/allennlp-as-a-library-example/tree/43b7dac8097a0c4a50c0b25594daa123f3889b75),
+we just add a `requirements.txt` file specifying `allennlp` as a dependency.  We referenced a
+particular commit on github, but you can just give a version number, like `allennlp==0.3`.  Then,
+after creating a python 3.6 environment, you install AllenNLP by running `pip install -r
+requirements.txt`.
```
Comment (joelgrus) [inline tutorials/getting_started/another_model_example.md]: with our current setup don't you also have to install pytorch and the spacy model?

File: tutorials/getting_started/another_model_example.md [hunk 6]
```diff
@@ -0,0 +1,611 @@
...
+from my_library.dataset_readers import SemanticScholarDatasetReader
+
+class TestSemanticScholarDatasetReader(AllenNlpTestCase):
+    def test_read_from_file(self):
+        reader = SemanticScholarDatasetReader()
+        dataset = reader.read('tests/fixtures/s2_papers.jsonl')
+```
+
+Then we just want to make sure that the resulting dataset looks like we expect.  We'll refer you
+to the [dataset tutorial](TODO - how do I link this??) for a deeper dive on the `Dataset`,
```
Comment (joelgrus) [inline tutorials/getting_started/another_model_example.md]: you can just link to the slug in the tutorials url. that tutorial is at http://allennlp.org/tutorials/data-pipeline

so you should be able to just do `[dataset tutorial](data-pipeline)`

Comment (joelgrus) [reply tutorials/getting_started/another_model_example.md]: (that makes it work on the website but not in the github repo, unfortunately)

File: tutorials/getting_started/another_model_example.md [hunk 7]
```diff
@@ -0,0 +1,611 @@
...
+        output_dict['label'] = labels
+        return output_dict
+```
+
+The name here is a little misleading; originally, we were using this for doing things like viterbi
+decoding or beam search for more complex models, but we now do that in `forward`, typically.
+`decode` is currently just used for making a model's output human-readable (e.g., for the demo).
+We use the vocabulary and translate ids that we see to strings.
+
+Alright, we've got ourselves a model.  We can run the tests again using `pytest`, and confirm that
```
Comment (joelgrus) [inline tutorials/getting_started/another_model_example.md]: you've used this "Alright," a couple of times

File: tutorials/getting_started/another_model_example.md (hunk 5 above)
Comment (matt-gardner) [reply tutorials/getting_started/another_model_example.md]: Right, thanks.

File: tutorials/getting_started/another_model_example.md (hunk 6 above)
Comment (matt-gardner) [reply tutorials/getting_started/another_model_example.md]: Thanks.

File: tutorials/getting_started/another_model_example.md (hunk 3 above)
Comment (matt-gardner) [reply tutorials/getting_started/another_model_example.md]: Yeah, you're right.  Fixed.

File: tutorials/getting_started/another_model_example.md (hunk 7 above)
Comment (matt-gardner) [reply tutorials/getting_started/another_model_example.md]: Yeah, it was very late as I was writing this, and I didn't ever give it another read-through.  I'll go over it again to fix some of these wording issues.

File: tutorials/getting_started/another_model_example.md (hunk 2 above)
Comment (matt-gardner) [reply tutorials/getting_started/another_model_example.md]: covfifith...

File: tutorials/getting_started/another_model_example.md (hunk 1 above)
Comment (matt-gardner) [reply tutorials/getting_started/another_model_example.md]: Good point =).  Getting fatter by the day...

**Comment by matt-gardner** [comment]
Ok, I've fixed all the things you mentioned.  My remaining questions are around where to put this tutorial and what to call it.  I'm super happy with "Another model tutorial", but I'm not sure what else to do.

**Comment by joelgrus** [comment]
I would probably call it "using AllenNLP as a library" or similar

File: tutorials/getting_started/another_model_example.md [hunk 8]
```diff
@@ -0,0 +1,606 @@
...
+the `from_params` method.
+
+```python
+@DatasetReader.register("s2_papers")
+class SemanticScholarDatasetReader(DatasetReader):
+    def __init__(self,
+                 tokenizer: Tokenizer = None,
+                 token_indexers: Dict[str, TokenIndexer] = None) -> None:
+        self._tokenizer = tokenizer or WordTokenizer()
+        self._token_indexers = token_indexers or {"tokens": SingleIdTokenIndexer()}
```
Comment (schmmd) [inline tutorials/getting_started/another_model_example.md]: You could mention what `SingleIdTokenIndexer` does.

Comment (matt-gardner) [reply tutorials/getting_started/another_model_example.md]: The text currently says,

```
The constructor takes these dependencies as inputs, where by default we're splitting text into  
words and representing them as single word ids (as opposed to, say, sequences of character ids).
```

Is that not enough, given the description in the paragraph above?  How would you word this?

**Comment by matt-gardner** [comment]
@joelgrus, I named it "Using AllenNLP in Your Repo", and re-did the first few paragraphs.  Look ok?

Though, now that I think about it, "repo" might not be obvious to some users...

**Comment by schmmd** [comment]
Looks great--very detailed! I started writing comments about potentially defining some components (such as , but ended up deciding there was plenty of detail ;-)

**Comment by joelgrus** [comment]
I might say "Using AllenNLP in Your Project"

**Comment by matt-gardner** [comment]
Is everyone ok with this now, then?  Should I merge it?  No one has actually approved this yet... =).

**Comment by maxdel** [comment]
Actually, I recently studied tutorial's GitHub repo and used it in my project. My little feedback is that a note on prediction functionality would be helpful (inherit Predictor, register it under the model's name, etc).

And thanks for what you do.

**Comment by matt-gardner** [comment]
Thanks, that's a great point!  We need a tutorial on using your model with `predict`, and with the demo.

**Pull request merged**